# Sephora

In [2]:
import pandas as pd
import numpy as np

In [4]:
products=pd.read_csv("product_info.csv")
reviews = pd.read_csv("reviews_0-250.csv", low_memory=False)

In [6]:
products.shape

(8494, 27)

In [8]:
reviews.shape

(602130, 19)

In [9]:
products.head()

,product_id,product_name,brand_id,brand_name,loves_count,rating,reviews,size,variation_type,variation_value,...,online_only,out_of_stock,sephora_exclusive,highlights,primary_category,secondary_category,tertiary_category,child_count,child_max_price,child_min_price
0,P473671,Fragrance Discovery Set,6342,19-69,6320,3.6364,11.0,NaN,NaN,NaN,...,1,0,0,"['Unisex/ Genderless Scent', 'Warm &Spicy Scen...",Fragrance,Value & Gift Sets,Perfume Gift Sets,0,NaN,NaN
1,P473668,La Habana Eau de Parfum,6342,19-69,3827,4.1538,13.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,85.0,30.0
2,P473662,Rainbow Bar Eau de Parfum,6342,19-69,3253,4.2500,16.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
3,P473660,Kasbah Eau de Parfum,6342,19-69,3018,4.4762,21.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
4,P473658,Purple Haze Eau de Parfum,6342,19-69,2691,3.2308,13.0,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0


In [11]:
products.isnull().sum()

product_id               0
product_name             0
brand_id                 0
brand_name               0
loves_count              0
rating                 278
reviews                278
size                  1631
variation_type        1444
variation_value       1598
variation_desc        7244
ingredients            945
price_usd                0
value_price_usd       8043
sale_price_usd        8224
limited_edition          0
new                      0
online_only              0
out_of_stock             0
sephora_exclusive        0
highlights            2207
primary_category         0
secondary_category       8
tertiary_category      990
child_count              0
child_max_price       5740
child_min_price       5740
dtype: int64

In [12]:
reviews.isnull().sum()

Unnamed: 0                       0
author_id                        0
rating                           0
is_recommended              117486
helpfulness                 331832
total_feedback_count             0
total_neg_feedback_count         0
total_pos_feedback_count         0
submission_time                  0
review_text                    999
review_title                167011
skin_tone                   106056
eye_color                   138488
skin_type                    74683
hair_color                  141081
product_id                       0
product_name                     0
brand_name                       0
price_usd                        0
dtype: int64

# Data Cleaning

In [13]:
# Clean products
products_clean = products.drop(columns=['value_price_usd', 'sale_price_usd', 'child_max_price', 'child_min_price'])
products_clean = products_clean.dropna(subset=['rating', 'reviews'])
products_clean['variation_type'] = products_clean['variation_type'].fillna('Not specified')
products_clean['variation_value'] = products_clean['variation_value'].fillna('Not specified')
products_clean['variation_desc'] = products_clean['variation_desc'].fillna('Not specified')
products_clean['size'] = products_clean['size'].fillna('Not specified')
products_clean['ingredients'] = products_clean['ingredients'].fillna('Unknown')
products_clean['highlights'] = products_clean['highlights'].fillna('Unknown')
products_clean['secondary_category'] = products_clean['secondary_category'].fillna('Uncategorized')
products_clean['tertiary_category'] = products_clean['tertiary_category'].fillna('Uncategorized')

print("Products clean shape:", products_clean.shape)
products_clean.isnull().sum()

Products clean shape: (8216, 23)


product_id            0
product_name          0
brand_id              0
brand_name            0
loves_count           0
rating                0
reviews               0
size                  0
variation_type        0
variation_value       0
variation_desc        0
ingredients           0
price_usd             0
limited_edition       0
new                   0
online_only           0
out_of_stock          0
sephora_exclusive     0
highlights            0
primary_category      0
secondary_category    0
tertiary_category     0
child_count           0
dtype: int64

In [14]:
# Clean reviews
reviews_clean = reviews.drop(columns=['Unnamed: 0', 'helpfulness'])
reviews_clean = reviews_clean.dropna(subset=['review_text'])
reviews_clean['is_recommended'] = reviews_clean['is_recommended'].fillna(-1)
reviews_clean['review_title'] = reviews_clean['review_title'].fillna('Unknown')
reviews_clean['skin_tone'] = reviews_clean['skin_tone'].fillna('Unknown')
reviews_clean['eye_color'] = reviews_clean['eye_color'].fillna('Unknown')
reviews_clean['skin_type'] = reviews_clean['skin_type'].fillna('Unknown')
reviews_clean['hair_color'] = reviews_clean['hair_color'].fillna('Unknown')

print("Reviews clean shape:", reviews_clean.shape)
reviews_clean.isnull().sum()

Reviews clean shape: (601131, 17)


author_id                   0
rating                      0
is_recommended              0
total_feedback_count        0
total_neg_feedback_count    0
total_pos_feedback_count    0
submission_time             0
review_text                 0
review_title                0
skin_tone                   0
eye_color                   0
skin_type                   0
hair_color                  0
product_id                  0
product_name                0
brand_name                  0
price_usd                   0
dtype: int64

## Data Cleaning Report

Started with 8,494 products and 602,130 reviews (using the reviews_0-250 chunk — a solid 600K+ sample instead of pulling all ~1M rows, since that would just add processing time without changing the actual insights).

**Products:**
- Dropped `value_price_usd`, `sale_price_usd`, `child_max_price`, `child_min_price` — these were 68-97% empty, not usable for real analysis.
- Dropped 278 rows with no rating — a product with no rating means it has no reviews yet, so it's not useful for anything I'm analyzing here.
- Filled in variation/size/ingredients/highlights/category gaps with "Not specified" / "Unknown" / "Uncategorized" instead of dropping those rows — didn't want to lose otherwise good product data over one missing field.
- Final: 8,216 products, no nulls left.

**Reviews:**
- Dropped the leftover index column and `helpfulness` (55%+ missing, not reliable enough to use).
- Dropped 999 rows with no review text — no text means no sentiment analysis possible for that row.
- Kept `is_recommended` but filled missing values with -1 (flagging "not specified") instead of dropping 117K+ rows — too much data to lose.
- Filled missing skin_tone/eye_color/skin_type/hair_color/review_title with "Unknown" — still useful for segmentation on the rows that do have this info.
- Final: 601,131 reviews, no nulls left.

**Note on scope:** This analysis uses the reviews_0-250 file only, which covers a product subset rather than all 8,216 products. Product-level analysis (pricing, ratings, categories) covers the full catalog. Review-based analysis (sentiment, skin type/tone patterns) reflects this subset — a reasonable trade-off for keeping the project focused instead of processing 1M+ rows for no real analytical gain.

In [15]:
products_clean.to_csv("products_clean.csv", index=False)
reviews_clean.to_csv("reviews_clean.csv", index=False)
print("Exported both clean CSVs")

Exported both clean CSVs


In [16]:
!pip install psycopg2-binary sqlalchemy

Defaulting to user installation because normal site-packages is not writeable


In [20]:
from sqlalchemy import create_engine, text
from urllib.parse import quote_plus

password = quote_plus("Arya@980")
engine = create_engine(f"postgresql://postgres:{password}@localhost:5432/sephora_analytics")

# actually test the connection this time
with engine.connect() as conn:
    result = conn.execute(text("SELECT 1"))
    print(result.fetchone())

(1,)


In [21]:
products_clean.to_sql("products", engine, if_exists="replace", index=False)
reviews_clean.to_sql("reviews", engine, if_exists="replace", index=False)
print("Pushed both tables")

Pushed both tables
